Updated docs for ROBO support

stuclem opened the pull request:
Fixes https://github.com/vmware/vic-product/issues/1554. Adds vSphere ROBO Advanced as a supported vSphere edition, and updates the DRS requirements for deployment to clusters.

@cgtexmex does this cover it, or do we need to add more? Thanks!

stuclem commented:
@cgtexmex do we need to add permissions for folders to [Manually Create a User Account for the Operations User](https://vmware.github.io/vic-product/assets/files/html/1.4/vic_vsphere_admin/ops_user_manual.html)?

stuclem commented:
https://github.com/vmware/vic-product/pull/1617/commits/d2b91a233dfb48e46ebd7f81199b8919c1c3df9d fixes https://github.com/vmware/vic-product/issues/1604, to move the global permissions for the ops user from `VCH - endpoint` to `VCH - vcenter`. @anchal-agrawal is this OK? Is this all that needs to be done for the manual configuration of ops user in a non-DRS environment? Thanks!

stuclem commented:
Towards https://github.com/vmware/vic-product/issues/1562 (verify/update installation prereqs).

cgtexmex commented on docs/user_doc/vic_overview/intro_to_vic.md:
```diff
@@ -15,7 +15,10 @@ vSphere Integrated Containers is delivered as an appliance, that comprises the f
 - **vSphere Integrated Containers Management Portal**, a container management portal that provides a UI for DevOps teams to provision and manage containers, including the ability to obtain statistics and information about container instances. Cloud administrators can manage container hosts and apply governance to their usage, including capacity quotas and approval workflows. Cloud administrators  ...
 
-These components currently support the Docker image format. vSphere Integrated Containers is entirely Open Source and free to use. Support for vSphere Integrated Containers is included in the vSphere Enterprise Plus license.
+These components currently support the Docker image format. vSphere Integrated Containers is entirely Open Source and free to use. Support for vSphere Integrated Containers is included with the following vSphere Editions:
```
>Support for vSphere Integrated Containers is included with the following vSphere Editions:

Should we prefix with "Full" (like below) or "Global Services"?  It is OSS that means we'll help (as best we can) free users....unless I'm missing something...

cgtexmex replied:
Possibly that's a moot point -- the reality is the features that vic require are only available in a Plus or ROBO Advanced license...which you hit in Installation...

stuclem replied:
Good point. How's this? "vSphere Integrated Containers depends on certain features that are included in the following vSphere Editions:"

cgtexmex replied:
very nice..

cgtexmex commented on docs/user_doc/vic_vsphere_admin/interop.md:
```diff
@@ -27,7 +27,7 @@ Your organization could use VMware vRealize Automation to provide a self-provisi
 ## VMware vSphere vMotion <a id="vmotion"></a>
 
-You can use vMotion to move VCHs without needing to take the container VMs offline. The VCH endpoint VM does not need to be running for vMotion to occur on the container VMs. Clusters with a mix of container VMs and non-container VMs can use vMotion with fully automated DRS. 
+You can use vMotion to move VCHs without needing to take the container VMs offline. The VCH endpoint VM does not need to be running for vMotion to occur on the container VMs. Clusters with a mix of container VMs and non-container VMs can use vMotion with fully automated VMware vSphere Distributed Resource Scheduler (DRS). 
```
>Clusters with a mix of container VMs and non-container VMs can use vMotion with fully automated VMware vSphere Distributed Resource Scheduler (DRS).

DRS Enablement isn't a requirement for vMotion use.....vic will work in a DRS Enabled environment just like any other VM...

stuclem replied:
Replaced with "You can use vMotion on clusters with a mix of container VMs and non-container VMs."

cgtexmex replied:
bravo

cgtexmex commented on docs/user_doc/vic_vsphere_admin/vch_compute.md:
```diff
@@ -29,7 +29,16 @@ The sections in this topic each correspond to an entry in the Compute Capacity p
+When deploying VCHs to a cluster, you cannot deploy a VCH to a specific host in the cluster. You deploy the VCH to the cluster and the placement of the VCH is determined as follows: 
+
+- If VMware vSphere Distributed Resource Scheduler (DRS) is enabled on that cluster, DRS manages the placement of the VCH on a host. 
+- If DRS is not enabled, vSphere Integrated Containers selects a host in the cluster on which to deploy the VCH. 
```
> If DRS is not enabled, vSphere Integrated Containers selects a host in the cluster on which to deploy the VCH. 

Not quite accurate -- we always trust the cluster to deploy the container on a usable host, vic will determine if that host is appropriate for powering on the container.  In a DRS enabled environment that power on decision is facilitated by DRS...so it's less about placement and more about powering on...but maybe a moot point...back to my earlier minutiae concern..

stuclem replied:
How's this?

The placement of the VCH, and its associated container VMs, is determined according to whether a given host is appropriate for powering on container VMs.

- If VMware vSphere Distributed Resource Scheduler (DRS) is enabled on that cluster, DRS manages the placement of the VCH and container VMs on hosts.
- If DRS is not enabled, the cluster selects a host and vSphere Integrated Containers checks whether the selected host is appropriate for the powering on of container VMs.

cgtexmex replied:
That sounds good...but I might tweak it to say:

> If DRS is not enabled, the cluster selects a host and vSphere Integrated Containers selects the host for powering on the container VMs.

The powerOn target may or may not be the host it was created on...

cgtexmex commented:
Overall I think this looks good -- some (possibly overly verbose) commentary...let me know if you have questions or concerns.

stuclem commented on docs/user_doc/vic_overview/intro_to_vic_engine.md:
```diff
@@ -49,11 +49,19 @@ A VCH is functionally distinct from a traditional container host in the followin
+
+VCHs are deployed as resource pools. The resource pool provides a useful visual parent-child relationship in the vSphere Client so that you can easily identify the container VMs that are provisioned into a VCH. You can also specify resource limits on the resource pool. You can provision multiple VCHs onto a single ESXi host, into a resource pool, or into a vCenter Server cluster.
+
+**NOTE**: Clusters that do not implement VMware vSphere Distributed Resource Scheduler (DRS) do not support resource pools. If you deploy a VCH to a cluster on which DRS is disabled, the VCH is created in a VM folder, rather than in a resource pool. This restricts your ability to configure resource usage limits on the VCH. It is strongly recommended that DRS is enabled on clusters to which you de ...
```
"VMware recommends that you enable DRS on clusters whenever possible." Is that better?

cgtexmex replied:
That gets my vote :)

stuclem commented on docs/user_doc/vic_overview/intro_to_vic_engine.md:
```diff
@@ -49,11 +49,19 @@ A VCH is functionally distinct from a traditional container host in the followin
 ## The VCH Endpoint VM <a id="endpoint"></a>
 
-The VCH endoint VM is the VM that runs inside the VCH resource pool. There is a 1:1 relationship between a VCH and a VCH endpoint VM. The VCH endpoint VM provides the following functions:
+The VCH endoint VM is the VM that runs inside the VCH resource pool or folder. There is a 1:1 relationship between a VCH and a VCH endpoint VM. The VCH endpoint VM provides the following functions:
```
Actually, I removed the note about non-DRS environments. It's too much info for the overview.

stuclem commented:
https://github.com/vmware/vic-product/pull/1617/commits/61216f1dd00fb2bbd37eca150141cbcd6dd6cbf7 fixes https://github.com/vmware/vic-product/issues/1598.

anchal-agrawal commented on docs/user_doc/vic_vsphere_admin/ops_user_manual.md:
```diff
@@ -29,7 +29,9 @@ Log into the Flex-based vSphere Web Client with a vSphere administrator account.
+  Global &gt; Disable methods</td>
 </tr>
 <tr>
 <td><code>VCH - datacenter</code></td>
```
In the automatic ops-user permissions grant code in vic-machine, we've reduced the `VCH - datacenter` role's set of privileges to only `Datastore > Configure datastore` and `Datastore > Low level file operations`. This is because the `VCH - endpoint` role is now _also_ applied to the VCH folder (`/datacenter/vm/my-vch`) so we can apply the `VCH - datacenter` role's `VirtualMachine` privileges at a more granular level (VCH folder vs. datacenter-level).

However, this presents an interesting scenario for setting ops-user permissions manually since the VCH folder is created by `vic-machine create` and cannot exist prior to running that command, i.e. the vSphere admin cannot manually pre-assign roles/permissions to the VCH folder. We have a few options here:

* Deprecate the docs for setting roles/permissions manually and advise the reader to use the `--ops-grant-perms` option in `vic-machine create`.
* Match the role/permission settings in `vic-machine` code to the manual settings here (this would be tricky since the VCH folder cannot pre-exist, and I'm not sure if `vic-machine create` will succeed if the VCH folder doesn't have the needed permissions)
* For setting roles/permissions manually, we list a feasible configuration which isn't the most granular but works if the vSphere admin wishes to set roles themselves. We can add a caveat that if the user wishes to have an ops-user configuration that assigns least privilege to entities and applies permissions at the most granular level, they should use the `--ops-grant-perms` option in `vic-machine create` instead of assigning roles/permissions manually.

I'm in favor of the latter and I'll verify whether it's possible.

anchal-agrawal replied:
@stuclem We'll need a separate section on this page for ROBO (non-DRS) environments. For a non-DRS environment, we use a new role (`VCH - endpoint - datastore`) that combines the privileges in the `VCH - datastore` and `VCH - endpoint` roles. The `VCH - endpoint - datastore` role is then applied to the cluster, in place of the `VCH - endpoint` role that's applied to the cluster in DRS environments.

anchal-agrawal replied:
> added a row for clusters in non-DRS environments, and applied both the datastore and endpoint roles to the clusters

Don't think it's possible to assign multiple roles to the same entity - please see https://github.com/vmware/vic/pull/7667#discussion_r180607660 and https://github.com/vmware/vic/issues/7721#issue-313842538. My initial code change assigned the two roles separately, but it was noted that assigning the second role undoes the first role's assignment. https://github.com/vmware/vic/pull/7726 fixed this by creating a single role that combines both sets of privileges.

> didn't mention ROBO, as this seems more like a DRS enabled/disabled issue

Agreed.

stuclem replied:
I see. I have attempted to remedy this in https://github.com/vmware/vic-product/pull/1617/commits/288b92475efc5d86dd5388e84551934bde3cb1aa. Is this clear enough?

anchal-agrawal commented:
@stuclem Unrelated to this change, but we should note in https://vmware.github.io/vic-product/assets/files/html/1.3/vic_vsphere_admin/set_up_ops_user.html that the `--ops-grant-perms` option in `vic-machine create` does not work when installing to a standalone ESX.

stuclem commented:
Added the `docker cp` privileges in https://github.com/vmware/vic-product/pull/1617/commits/8068d45ffd66c4e320bf1545ba6cef057c8fb379.

Added that `--ops-grant-perms` doesn't work on standalone hosts in https://github.com/vmware/vic-product/pull/1617/commits/e1a52709fb83fad2f3cc1ecc5ece13cc71262953.

anchal-agrawal commented on docs/user_doc/vic_vsphere_admin/ops_user_manual.md:
```diff
@@ -65,7 +71,9 @@ VirtualMachine &gt; Inventory &gt; Remove</td>
   VirtualMachine &gt; Configuration &gt; Remove disk<br>
   VirtualMachine &gt; Configuration &gt; Rename<br>
-  VirtualMachine &gt; Guest operations &gt; Guest operation program execution<br>
+  VirtualMachine`Guest operations &gt; Guest operation program execution<br>
```
This line renders as 
```
VirtualMachine`Guest operations > Guest operation program execution
```

anchal-agrawal commented on docs/user_doc/vic_vsphere_admin/ops_user_manual.md:
```diff
@@ -65,7 +71,9 @@ VirtualMachine &gt; Inventory &gt; Remove</td>
   VirtualMachine &gt; Configuration &gt; Rename<br>
-  VirtualMachine &gt; Guest operations &gt; Guest operation program execution<br>
+  VirtualMachine`Guest operations &gt; Guest operation program execution<br>
+  VirtualMachine &gt; GuestOperations &gt; Modify<br>
```
Minor: `Guest operations` instead of `GuestOperations`

anchal-agrawal commented on docs/user_doc/vic_vsphere_admin/ops_user_manual.md:
```diff
@@ -1,10 +1,16 @@
 # Manually Create a User Account for the Operations User #
 
-When you deploy a VCH, the user account that you specify as the operations user must have the correct privileges to allow the VCH to perform post-deployment operations. vSphere Integrated Containers Engine provides a mechanism to automatically assign the necessary permissions to the operations user account, but you can also choose to create the user account manually in vSphere. To do so, you crea ...
+When you deploy a virtual container host (VCH), the user account that you specify as the operations user must have the correct privileges to allow the VCH to perform post-deployment operations. vSphere Integrated Containers Engine provides a mechanism to automatically assign the necessary permissions to the operations user account. You can also create the user account manually in vSphere. 
```
> You can also create the user account manually in vSphere.

We should make it clear that regardless of whether the admin uses `vic-machine create --ops-grant-perms` or the manual method, the user account must already be created (manually or otherwise). So if they intend to use `vic-machine create --ops-grant-perms`, they must first create the user account in vCenter. `vic-machine create --ops-grant-perms` does not create the user - it only assign roles and permissions to it.

stuclem replied:
I think that we state this pretty clearly in the main topic, [Configure the Operations User > Create a User Account for the Operations User](https://vmware.github.io/vic-product/assets/files/html/1.4/vic_vsphere_admin/set_up_ops_user.html#createuser), but I restated it here too in the prerequisites. See https://github.com/vmware/vic-product/pull/1617/commits/6371ffb26d088eb6beaf3f54e3165023bdbfe3b7.

anchal-agrawal commented on docs/user_doc/vic_vsphere_admin/ops_user_manual.md:
```diff
@@ -1,10 +1,16 @@
 - For information about how to create vSphere roles, see [vSphere Permissions and User Management Tasks](https://docs.vmware.com/en/VMware-vSphere/6.7/com.vmware.vsphere.security.doc/GUID-5372F580-5C23-4E9C-8A4E-EF1B4DD9033E.html) in the vSphere documentation. 
 - For information about how to assign permissions to objects in the vSphere Inventory, see [Add a Permission to an Inventory Object](https://docs.vmware.com/en/VMware-vSphere/6.7/com.vmware.vsphere.security.doc/GUID-A0F6D9C2-CE72-4FE5-BAFC-309CFC519EC8.html) in the vSphere documentation.
 
+When creating roles manually, the privileges are not as granular as when you use the option to grant permissions automatically. VMware recommends that you use the option to grant permissions automatically whenever possible.
```
👍

stuclem commented on docs/user_doc/vic_vsphere_admin/ops_user_manual.md:
```diff
@@ -65,7 +71,9 @@ VirtualMachine &gt; Inventory &gt; Remove</td>
-  VirtualMachine &gt; Guest operations &gt; Guest operation program execution<br>
+  VirtualMachine`Guest operations &gt; Guest operation program execution<br>
+  VirtualMachine &gt; GuestOperations &gt; Modify<br>
+  VirtualMachine &gt; GuestOperations &gt; Query<br>
```
Fixed in https://github.com/vmware/vic-product/pull/1617/commits/6371ffb26d088eb6beaf3f54e3165023bdbfe3b7.

anchal-agrawal commented on docs/user_doc/vic_vsphere_admin/ops_user_manual.md:
```diff
@@ -43,20 +54,23 @@ VirtualMachine &gt; Inventory &gt; Create new<br>
 <tr>
-<td><code>VCH - datastore</code></td>
-<td>Datastore &gt; AllocateSpace<br>Datastore &gt; Browse datastore <br>Datastore &gt; Configure datastore<br>Datastore &gt; Remove file<br>Datastore &gt; Low level file operations<br>Host &gt; Configuration &gt; System management</td>
+<td><code>VCH - datastore</code><br /><br/>This role only applies to DRS environments. See <a href="#drsnote">note</a>.</td>
```
> This role only applies to DRS environments.

This is inaccurate - we apply the datastore role to datastores regardless of the DRS setting (https://github.com/vmware/vic/blob/master/lib/install/opsuser/opsuser.go#L219).

stuclem replied:
Fixed in https://github.com/vmware/vic-product/pull/1617/commits/fa9e210336f4b0a8c3cc1d406747eb4fa714141f. Thanks @anchal-agrawal!

anchal-agrawal commented on docs/user_doc/vic_vsphere_admin/ops_user_manual.md:
```diff
@@ -43,20 +54,23 @@ VirtualMachine &gt; Inventory &gt; Create new<br>
 </tr>
 <tr>
-<td><code>VCH - endpoint</code></td>
+<td><code>VCH - endpoint</code><br /><br/>This role only applies to DRS environments. See <a href="#drsnote">note</a>.</td>
```
> This role only applies to DRS environments.

This is inaccurate - we apply the endpoint role to the VCH folder regardless of the DRS setting (https://github.com/vmware/vic/blob/master/lib/install/opsuser/opsuser.go#L219).

stuclem replied:
Unchanged.

stuclem commented on docs/user_doc/vic_vsphere_admin/ops_user_manual.md:
```diff
@@ -66,24 +80,59 @@ VirtualMachine &gt; Inventory &gt; Remove</td>
+  dvPort group &gt; Modify<br>
+  dvPort group &gt; Policy operation<br>
+  dvPort group &gt; Scope operation<br>
+  Resource &gt; Assign virtual machine to resource pool<br>
```
Fixed in https://github.com/vmware/vic-product/pull/1617/commits/fa9e210336f4b0a8c3cc1d406747eb4fa714141f. Thanks @anchal-agrawal!
